sys/tsrb/tsrb.c

```diff
@@ -19,27 +19,27 @@
 
 #include "tsrb.h"
 
-static void _push(tsrb_t *rb, char c)
+static void _push(tsrb_t *rb, uint8_t c)
 {
     rb->buf[rb->writes++ & (rb->size - 1)] = c;
 }
 
-static char _pop(tsrb_t *rb)
+static uint8_t _pop(tsrb_t *rb)
 {
     return rb->buf[rb->reads++ & (rb->size - 1)];
 }
 
 int tsrb_get_one(tsrb_t *rb)
```

Comment:
why not change this to `uint8_t tsrb_get_one(tsrb_t *rb)` and remove the `-1` return, but change the documentation so that it states that the callee of this function SHOULD (or MUST) check `if (!tsrb_empty(rb))` before usage. 

That way we avoid necessary casts of `int` to `uint8_t` and its safe because this whole thing assumes single produces + single consumer.

Reply:
I don't think having such central function *not* return an error if used incorrectly is not a good idea.

Reply:
Also it's the same reason, why I disagree with a return pointer: it makes the function less efficient, especially when implementing such central functions like `getchar()`.

Reply:
why is it less efficient? Its simpler and (more over) type safe:

```
uint8_t tsrb_get_one(tsrb_t *rb)
{
    return _pop(rb);
}
```

and upon usage (like in isrpipe or so)

```
uint8_t c;
if (!tsrb_empty(rb)) {
    c = tsrb_get_one(rb);
}
else {
    DEBUG("ring buffer empty!\n");
    /* do error handling if required */
}
```

Reply:
Because instead of checking it once in the function, every use of `tsrb_get_one()` needs to add a check (so adding ROM for every call). This makes `tsrb_get_one()` maybe type-safe (which is due to the new typing arguable, as values and error codes are now two distinct value-spaces), but inherently error-prone in its usage.

Reply:
I don't think this is a good idea. _pop()ing too far completely messes up the ringbuffer. I'd rather not expose that towards the users.

All sane usages would look like your example. So IMO it is more sane to keep that as:

```
int c = tsrb_get_one(rb);
if (c < 0) {
     DEBUG("ring buffer empty!\n");
    /* do error handling if required */
}
```

Reply:
@kaspar030: oversaw the increment of `reads` so one MUST check if ring buffer not empty before any `get` operation. But then we could actually simplify the API by removing `tsrb_empty` and `tsrb_full` because they have no value to the user as s/he can just call `tsrb_get*` or `tsrb_add*` and check for errors.

Reply:
> But then we could actually simplify the API by removing `tsrb_empty` and `tsrb_full` because they have no value to the user as s/he can just call `tsrb_get*` or `tsrb_add*` and check for errors.

Different topic.

Reply:
> But then we could actually simplify the API by removing `tsrb_empty` and `tsrb_full` because they have no value to the user as s/he can just call `tsrb_get*` or `tsrb_add*` and check for errors.

Sometimes it is nice to check for full / not full *before* getting the data.
E.g., in an ISR handler that *can* read data from a register, but doesn't have to.

```
uint8_t c = read_and_delete_from_register();
if (tsrb_add(rb, c) < 0) {
  // now what, store "c" to global variable?
}
```

v.s.

```
if (!tsrb_full()) {
  tsrb_add(rb, read_and_delete_from_register();
}
```